crates/telio-traversal/src/last_handshake_time_provider.rs

```diff
@@ -0,0 +1,65 @@
+use std::time::Duration;
+
+use async_trait::async_trait;
+use telio_crypto::PublicKey;
+use telio_wg::{DynamicWg, WireGuard};
+
+pub const MAX_PEER_HANDSHAKE_TIMEOUT: Duration = Duration::from_secs(180);
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Unknown peer: {0}")]
+    UnknownPeer(PublicKey),
+    #[error("Uknown last handshake time for {0}")]
+    UnknownLastHandshakeTime(PublicKey),
+    #[error("Unable to communicate with wireguard: {0}")]
+    UnableToCommunicateWithWG(#[from] telio_wg::Error),
+}
+
+#[cfg_attr(any(test, feature = "mockall"), mockall::automock)]
+#[async_trait]
+pub trait LastHandshakeTimeProvider: Send + Sync {
+    /// Return last handshake time for a peer using given public key
+    async fn last_handshake_time(&self, public_key: &PublicKey) -> Result<Duration, Error>;
+}
+
+#[async_trait]
+impl LastHandshakeTimeProvider for DynamicWg {
+    async fn last_handshake_time(&self, public_key: &PublicKey) -> Result<Duration, Error> {
+        let interface = self.get_interface().await?;
+        match interface
+            .peers
+            .get(public_key)
+            .map(|peer| peer.time_since_last_handshake)
+        {
+            Some(Some(t)) => Ok(t),
+            Some(None) => Err(Error::UnknownLastHandshakeTime(*public_key)),
```

Comment:
I would recommend using `Result<Option<Duration>, Error>` as a return type. It is possible, that no handshake occured at all (e.g. we have never been connected to the peer), thus the `Duration` would be `None` in that case. It seems like it would represent the situation more naturally, when compared to `Err(...)`.

Reply:
done